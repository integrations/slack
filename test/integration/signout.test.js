const request = require('supertest');
const nock = require('nock');

const { probot, slackbot, models } = require('.');
const fixtures = require('../fixtures');

const promptUrl = /^http:\/\/127\.0\.0\.1:\d+(\/github\/oauth\/login\?state=(.*))/;

const {
  SlackWorkspace,
  GitHubUser,
  SlackUser,
  Installation,
  Subscription,
} = models;

describe('Integration: signout', async () => {
  let workspace;
  let githubUser;
  let slackUser;
  beforeEach(async () => {
    // create workspace
    workspace = await SlackWorkspace.create({
      slackId: 'T0001',
      accessToken: 'xoxp-token',
    });

    githubUser = await GitHubUser.create({
      id: 1,
      accessToken: 'secret',
    });

    slackUser = await SlackUser.create({
      slackId: 'U2147483697',
      slackWorkspaceId: workspace.id,
    });

    const installation1 = await Installation.create({
      githubId: 1,
      ownerId: 1,
    });

    const installation2 = await Installation.create({
      githubId: 2,
      ownerId: 2,
    });

    await Subscription.subscribe({
      channelId: 'C2147483705',
      githubId: 1,
      installationId: installation1.id,
      slackWorkspaceId: workspace.id,
      creatorId: slackUser.id,
      type: 'repo',
    });

    await Subscription.subscribe({
      channelId: 'C2147483705',
      githubId: 2,
      installationId: installation2.id,
      slackWorkspaceId: workspace.id,
      creatorId: slackUser.id,
      type: 'repo',
    });

    await Subscription.subscribe({
      channelId: 'C12345',
      githubId: 3,
      installationId: installation2.id,
      slackWorkspaceId: workspace.id,
      creatorId: slackUser.id,
      type: 'account',
    });
  });
  test('a signed out user is prompted to sign in first', async () => {
    const command = fixtures.slack.command({
      text: 'signout',
    });
    const res = await request(probot.server).post('/slack/command')
      .use(slackbot)
      .send(command)
      .expect(200);

    const { url } = res.body.attachments[0].actions[0];
    const [, link, state] = url.match(promptUrl);

    const loginRequest = request(probot.server).get(link);
    await loginRequest.expect(302).expect(
      'Location',
      `https://github.com/login/oauth/authorize?client_id=&state=${state}`,
    );

    // GitHub redirects back, authenticates user and process subscription
    nock('https://github.com').post('/login/oauth/access_token')
      .reply(200, fixtures.github.oauth);
    nock('https://api.github.com').get('/user')
      .reply(200, fixtures.user);

    nock('https://hooks.slack.com').post('/commands/1234/5678', (body) => {
      expect(body).toMatchSnapshot();
      return true;
    }).reply(200);

    await request(probot.server).get('/github/oauth/callback').query({ state })
      .expect(302)
      .expect(
        'Location',
        `https://slack.com/app_redirect?team=${command.team_id}&channel=${command.channel_id}`,
      );
  });

  test('a signed in user is successfully signed out and relevant subscriptions are deleted', async () => {
    await slackUser.update({
      githubId: githubUser.id,
    });
    nock('https://api.github.com').get('/repositories/1').reply(200, {
      full_name: 'atom/atom',
    });

    nock('https://api.github.com').get('/repositories/2').reply(200, {
      full_name: 'kubernetes/kubernetes',
    });

    nock('https://api.github.com').get('/user/3').reply(200, {
      login: 'github',
    });

    nock('https://slack.com').post('/api/chat.postMessage', (body) => {
      expect(body).toMatchSnapshot();
      return true;
    }).times(2).reply(200, { ok: true });

    const command = fixtures.slack.command({
      text: 'signout',
    });

    const res = await request(probot.server).post('/slack/command').send(command).expect(200);
    expect(res.body.attachments[0].text).toMatchSnapshot();

    expect((await slackUser.reload()).githubId).toBe(null);
    expect((await Subscription.findAll({ where: { creatorId: slackUser.id } })).length).toBe(0);
  });
});
