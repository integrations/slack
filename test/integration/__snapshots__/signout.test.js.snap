// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration: signout a signed in user is successfully signed out and relevant subscriptions are deleted 1`] = `
Object {
  "attachments": "[{\\"color\\":\\"#24292f\\",\\"mrkdwn_in\\":[\\"text\\"],\\"text\\":\\"Subscriptions to 2 repositories have been disabled because <@U2147483697>, who originally set them up, has disconnected their GitHub account.\\\\nUse the following slash commands to re-enable the subscriptions:\\\\n/github subscribe atom/atom\\\\n/github subscribe kubernetes/kubernetes\\"}]",
  "channel": "C2147483705",
  "token": "xoxp-token",
}
`;

exports[`Integration: signout a signed in user is successfully signed out and relevant subscriptions are deleted 2`] = `
Object {
  "attachments": "[{\\"color\\":\\"#24292f\\",\\"mrkdwn_in\\":[\\"text\\"],\\"text\\":\\"The subscription to 1 account has been disabled because <@U2147483697>, who originally set it up, has disconnected their GitHub account.\\\\nUse the following slash command to re-enable the subscription:\\\\n/github subscribe github\\"}]",
  "channel": "C12345",
  "token": "xoxp-token",
}
`;

exports[`Integration: signout a signed in user is successfully signed out and relevant subscriptions are deleted 3`] = `
":white_check_mark: <@U2147483697> is now signed out
Features like subscriptions and rich link previews will stop working. Use \`/github signin\` to sign back into your GitHub account at any time."
`;

exports[`Integration: signout a signed in user is successfully signed out and relevant subscriptions are deleted 4`] = `
Array [
  "pending-command#13345224609.738474920.8088930838d88f008e0",
]
`;

exports[`Integration: signout a signed in user is successfully signed out and relevant subscriptions are deleted 5`] = `
Array [
  Array [
    "pending-command#13345224609.738474920.8088930838d88f008e0",
    Object {
      "channel_id": "C2147483705",
      "channel_name": "test",
      "command": "/github",
      "enterprise_id": "E0001",
      "enterprise_name": "Globular%20Construct%20Inc",
      "response_url": "https://hooks.slack.com/commands/1234/5678",
      "team_domain": "example",
      "team_id": "T0001",
      "text": "signout",
      "token": "secret",
      "trigger_id": "13345224609.738474920.8088930838d88f008e0",
      "user_id": "U2147483697",
      "user_name": "Steve",
    },
  ],
]
`;

exports[`Integration: signout a signed out user is prompted to sign in first 1`] = `
Object {
  "channelSlackId": "C2147483705",
  "exp": Any<Number>,
  "githubOAuthState": Any<String>,
  "iat": Any<Number>,
  "replaySlashCommand": false,
  "teamSlackId": "T0001",
  "trigger_id": "13345224609.738474920.8088930838d88f008e0",
  "userSlackId": "U2147483697",
}
`;

exports[`Integration: signout a signed out user is prompted to sign in first 2`] = `
Array [
  Object {
    "color": "#24292f",
    "mrkdwn_in": Array [
      "text",
    ],
    "text": ":white_check_mark: Success! <@U2147483697> is now connected to <https://github.com/wilhelmklopp|@wilhelmklopp>",
  },
]
`;

exports[`Integration: signout a signed out user is prompted to sign in first 3`] = `
Array [
  "pending-command#13345224609.738474920.8088930838d88f008e0",
]
`;

exports[`Integration: signout a signed out user is prompted to sign in first 4`] = `
Array [
  Array [
    "pending-command#13345224609.738474920.8088930838d88f008e0",
    Object {
      "channel_id": "C2147483705",
      "channel_name": "test",
      "command": "/github",
      "enterprise_id": "E0001",
      "enterprise_name": "Globular%20Construct%20Inc",
      "response_url": "https://hooks.slack.com/commands/1234/5678",
      "team_domain": "example",
      "team_id": "T0001",
      "text": "signout",
      "token": "secret",
      "trigger_id": "13345224609.738474920.8088930838d88f008e0",
      "user_id": "U2147483697",
      "user_name": "Steve",
    },
  ],
]
`;
